File: api/src/modules/_shared/whatsapp-hooks/infrastructure/mappers/whatsapp-message.mapper.ts
import type { WhatsappMessage } from "../../../../../lib/generated/prisma/client.js";
import type { IWhatsappMessage } from "../../domain/entities/whatsapp-message.entity.js";

export function toWhatsappMessageEntity(data: WhatsappMessage): IWhatsappMessage {
  return {
    id: data.id,
    webhook_log_id: data.webhookLogId,

    chat_id: data.chatId,
    chat_lid: data.chatLid,
    from: data.from,
    from_lid: data.fromLid,
    from_name: data.fromName,
    is_from_me: data.isFromMe,
    wa_timestamp: data.waTimestamp,

    message_type: data.messageType,
    body: data.body,

    replied_to_id: data.repliedToId,
    quoted_body: data.quotedBody,
    is_forwarded: data.isForwarded,

    media_path: data.mediaPath,
    full_media_path: data.mediaPath
      ? `${(process.env.APP_BASE_URL || "").endsWith("/") ? process.env.APP_BASE_URL : (process.env.APP_BASE_URL || "") + "/"}${data.mediaPath.startsWith("/") ? data.mediaPath.substring(1) : data.mediaPath}`
      : null,
    media_caption: data.mediaCaption,
    original_url: data.originalUrl,

    latitude: data.latitude,
    longitude: data.longitude,
    location_thumbnail: data.locationThumbnail,
    location_sequence: data.locationSequence,

    contact_name: data.contactName,
    contact_vcard: data.contactVcard,
    contacts: data.contacts,

    reaction: data.reaction,
    reacted_message_id: data.reactedMessageId,

    created_at: data.createdAt,
    updated_at: data.updatedAt,
  };
}

File: handlerclaw/lib/features/whatsapp-hooks/data/mappers/whatsapp_message_mapper.dart
import 'package:handlerclaw/features/whatsapp-hooks/data/dto/whatsapp_message_dto.dart';
import 'package:handlerclaw/features/whatsapp-hooks/domain/entities/whatsapp_message_entity.dart';

class WhatsappMessageMapper {
  static WhatsappMessageEntity toEntity(WhatsappMessageDto dto) {
    return WhatsappMessageEntity(
      id: dto.id,
      webhookLogId: dto.webhookLogId,
      chatId: dto.chatId,
      chatLid: dto.chatLid,
      from: dto.from,
      fromLid: dto.fromLid,
      fromName: dto.fromName,
      isFromMe: dto.isFromMe,
      waTimestamp: dto.waTimestamp,
      messageType: WhatsappMessageType.fromString(dto.messageType),
      body: dto.body,
      repliedToId: dto.repliedToId,
      quotedBody: dto.quotedBody,
      isForwarded: dto.isForwarded,
      mediaPath: dto.mediaPath,
      fullPathMedia: dto.fullMediaPath,
      mediaCaption: dto.mediaCaption,
      originalUrl: dto.originalUrl,
      latitude: dto.latitude,
      longitude: dto.longitude,
      locationThumbnail: dto.locationThumbnail,
      locationSequence: dto.locationSequence != null
          ? BigInt.tryParse(dto.locationSequence!)
          : null,
      contactName: dto.contactName,
      contactVcard: dto.contactVcard,
      contacts: dto.contacts,
      reaction: dto.reaction,
      reactedMessageId: dto.reactedMessageId,
      createdAt: dto.createdAt,
      updatedAt: dto.updatedAt,
    );
  }
}
